[4.1] Add Visual Indicator for Failing Scheduled Tasks

ditsuke opened the pull request:
Pull Request for Issue #36459.

### Summary of Changes
- Adds an icon as a visual indicator for failing tasks.
- Includes an on-hover tooltip for accessibility (review).


### Testing Instructions
- Create a new scheduled task with routine GET Request, setting an invalid URL as the target
.
- Let it run. It will fail.


### Actual result BEFORE applying this Pull Request
There is no visual indicator that the task has failed in the default Scheduled tasks view.

![before-patch](https://i.ibb.co/WzLF7TW/image.png)

### Expected result AFTER applying this Pull Request
There is a visual indicaator that the task is failing:
![after-patch](https://i.ibb.co/P638CM7/image.png)



### Documentation Changes Required

None

ditsuke committed:
Add visual indicator for failing scheduled tasks

Fixes joomla#36459. Adds a visual indicator to failing tasks in the form
of a warning/danger icon colored yellow. Other color schemes were tested
but this seemed like the best option to me. Open to change.
The visual indicator includes a tooltip for accessibility (not sure if
it needs something else).

ditsuke committed:
Merge branch '4.1-dev' of https://github.com/joomla/joomla-cms into add-visual-indication-for-failing-tasks-in-manager-36459

brianteeman commented:
Please look at other "tooltips" and create this one in the same way.

```
	<span class="icon-info-circle" aria-hidden="true" tabindex="0"></span>
	<div role="tooltip" class="text-start" id="tip<?php echo $language->code; ?>">
		<?php echo Text::_('JGLOBAL_LANGUAGE_VERSION_NOT_PLATFORM'); ?>
	</div>

```

ditsuke commented:
/cc @PhilETaylor please review

richard67 commented:
@ditsuke Could you check the [previous comment](https://github.com/joomla/joomla-cms/pull/36765#issuecomment-1018281670)? Thanks in advance.

ditsuke committed:
Update indicator-tooltip for consistency

ditsuke committed:
Merge branch '4.1-dev' into add-visual-indication-for-failing-tasks-in-manager-36459

ditsuke committed:
Add aria-hidden to indicator for a11y

richard67 commented on administrator/components/com_scheduler/tmpl/tasks/default.php:
```diff
@@ -198,12 +202,20 @@ class="js-draggable" data-url="<?php echo $saveOrderingUrl; ?>" data-direction="
-							<?php endif; ?>
+							<span class="task-title">
+								<?php if ($canEdit): ?>
+									 <a href="<?php echo Route::_('index.php?option=com_scheduler&task=task.edit&id=' . $item->id); ?>"
+										 title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape($item->title); ?>"> <?php echo $this->escape($item->title); ?></a>
```
```suggestion
									<a href="<?php echo Route::_('index.php?option=com_scheduler&task=task.edit&id=' . $item->id); ?>"
										title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape($item->title); ?>"> <?php echo $this->escape($item->title); ?>
									</a>
```

richard67 replied:
Remove space before the anchor and use tabs and not spaced to indent.

richard67 replied:
@ditsuke Drone PHPCS doesn't complain about this one, but it should be fixed, too.

richard67 commented on administrator/components/com_scheduler/tmpl/tasks/default.php:
```diff
@@ -198,12 +202,20 @@ class="js-draggable" data-url="<?php echo $saveOrderingUrl; ?>" data-direction="
+								<?php else: ?>
+									 <?php echo $this->escape($item->title); ?>
+								<?php endif; ?>
+								<?php if (!in_array($item->last_exit_code, [Status::OK, Status::WILL_RESUME])): ?>
+									<span class="failure-indicator icon-solid icon-exclamation-triangle hasTooltip" aria-hidden="true"></span>
+									<div role="tooltip">
+        								<?php echo Text::sprintf("COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING", $item->last_exit_code); ?>
+    								</div>
+								<?php endif; ?>
```
```suggestion
								<?php if (!in_array($item->last_exit_code, [Status::OK, Status::WILL_RESUME])): ?>
									<span class="failure-indicator icon-solid icon-exclamation-triangle hasTooltip" aria-hidden="true"></span>
									<div role="tooltip">
										<?php echo Text::sprintf("COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING", $item->last_exit_code); ?>
									</div>
								<?php endif; ?>
```

richard67 replied:
Use consistent indent.

richard67 replied:
This will also fix drone, I think.

richard67 replied:
There were also spaces and not tabs.

ditsuke replied:
Thanks

ditsuke committed:
Fix codestyle

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

ditsuke committed:
Fix codestyle (2)

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

chmst commented:
I have tested this item :white_check_mark: successfully on 3febb572a01b0f52e381c148de4a7d3afc212f40<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

richard67 committed:
Merge branch '4.1-dev' into add-visual-indication-for-failing-tasks-in-manager-36459

heelc29 commented on administrator/components/com_scheduler/tmpl/tasks/default.php:
```diff
@@ -62,7 +63,10 @@
 $document->addScriptOptions('com_scheduler.test-task.token', Session::getFormToken());
-$document->getWebAssetManager()->useScript('com_scheduler.test-task');
+
+$wa = $document->getWebAssetManager();
+$wa->useScript('com_scheduler.test-task');
+$wa->useStyle('com_scheduler.admin-view-tasks-css');
```
```suggestion
/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
$wa = $this->document->getWebAssetManager();
$wa->useScript('com_scheduler.test-task')
	->useStyle('com_scheduler.admin-view-tasks-css');
```

heelc29 commented on administrator/components/com_scheduler/tmpl/tasks/default.php:
```diff
@@ -62,7 +63,10 @@
 
 $document = $app->getDocument();
 $document->addScriptOptions('com_scheduler.test-task.token', Session::getFormToken());
```
```suggestion
$this->document->addScriptOptions('com_scheduler.test-task.token', Session::getFormToken());
```

ditsuke replied:
@heelc29 thanks!

ditsuke committed:
Apply suggestions from code review

Minor improvements to code quality/contextuality.

Co-authored-by: heelc29 <66922325+heelc29@users.noreply.github.com>

ditsuke committed:
Merge branch '4.1-dev' into add-visual-indication-for-failing-tasks-in-manager-36459

pritam825 commented:
![Screenshot (1981)](https://user-images.githubusercontent.com/54195688/151766615-7c0c9a10-0b4e-431a-94eb-ced1d086c891.png)
After applying the patch it' giving me error with this -> "There is no "com_scheduler.admin-view-tasks-css" asset of a "style" type in the registry".

pritam825 commented:
I think you forgot to declare this -> "com_scheduler.admin-view-tasks-css"

richard67 commented:
@pritam825 Have you "compiled" the css with `npm ci`or `npm run build:css`as described here https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment after having applied the patch? This will copy the css from "build/media_source/com_scheduler/css" to "media/com_scheduler/css".

pritam825 commented:
> @pritam825 Have you "compiled" the css with `npm ci`or `npm run build:css`as described here https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment after having applied the patch? This will copy the css from "build/media_source/com_scheduler/css" to "media/com_scheduler/css".

sorry sir, I forgot to use npm run build:css, Thanks sir

pritam825 commented:
It's still giving me the same error

richard67 commented:
> It's still giving me the same error

@pritam825 Try `npm ci` instead of `npm run build:css` and clear browser cache before testing again.

pritam825 commented:
@richard67  thanks, now it's working fine.
Working as expected
![Screenshot (1984)](https://user-images.githubusercontent.com/54195688/151772559-70336bb6-71fa-40a2-b694-c24ef4f52932.png)

richard67 commented:
@pritam825 Then mark your test result in the issue tracker. Thanks in advance.

pritam825 commented:
ok sir, Thanks

pritam825 commented:
I have tested this item :white_check_mark: successfully on a9d6300100f69f67b5cf9d0fc9cd1157c0bafe55<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

richard67 commented:
@chmst Could you test again? Thanks in advance.

jwaisner commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

jwaisner commented:
Back to pending.

@ditsuke ,

 Can you please fix the conflicts?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

richard67 committed:
Merge branch '4.1-dev' into add-visual-indication-for-failing-tasks-in-manager-36459

richard67 commented:
@ditsuke @jwaisner I've allowed myself to fix the merge conflict and to add back the previous results since the change after those just were a clean branch update with the conflict fix. => RTC again.

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

ditsuke commented:
@richard67 @jwaisner thanks!

Quy commented on administrator/components/com_scheduler/tmpl/tasks/default.php:
```diff
@@ -199,12 +201,21 @@ class="js-draggable" data-url="<?php echo $saveOrderingUrl; ?>" data-direction="
+									 <?php echo $this->escape($item->title); ?>
+								<?php endif; ?>
+								<?php if (!in_array($item->last_exit_code, [Status::OK, Status::WILL_RESUME])): ?>
+									<span class="failure-indicator icon-solid icon-exclamation-triangle hasTooltip" aria-hidden="true"></span>
```
```suggestion
									<span class="failure-indicator icon-exclamation-triangle" aria-hidden="true"></span>
```

Quy commented on administrator/language/en-GB/com_scheduler.ini:
```diff
@@ -81,6 +81,7 @@ COM_SCHEDULER_LAST_RUN_DATE="Last Run Date"
 COM_SCHEDULER_MANAGER_TASKS="Scheduled Tasks"
 COM_SCHEDULER_MANAGER_TASK_EDIT="Edit Task"
 COM_SCHEDULER_MANAGER_TASK_NEW="New Task"
+COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING="Task is failing with exit code %1$d"
```
```suggestion
COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING="Task failed. Exit code: %1$d"
```

ditsuke committed:
Apply suggestions from code review

Removes some CSS classes from the failure indicator and improves a language string.

Co-authored-by: Quy <quy@fluxbb.org>

ditsuke commented:
@richard67 applied

richard67 commented:
I have tested this item :white_check_mark: successfully on 05dde691b766a0c2521e992b10ef638ddb5e3208<br /><br />I've restored the previous test results since the change after that was just removal of 2 unused CSS classes in the markup and an improvement on the language string for the tool tip.

In addition I have tested it, too, just to be sure.

=> RTS still valid.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36765">issues.joomla.org/tracker/joomla-cms/36765</a>.</sub>

bembelimen commented:
Thx

ditsuke commented:
Thanks!
